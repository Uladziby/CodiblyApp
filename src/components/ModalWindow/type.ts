/** @format */

import { ReactNode } from "react";

export interface ModalWindowProps {
	onCloseHandler?: () => void;
	children: ReactNode | string;
}
