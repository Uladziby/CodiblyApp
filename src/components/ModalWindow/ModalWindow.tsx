/** @format */

import { MouseEvent } from "react";
import ReactDOM from "react-dom";
import { CloseButton, StyledContainer, StyledCover, StyledHeader } from "./ModalWindow.styles";
import { ReactComponent as IconClose } from "../../assets/clear.svg";
import { ModalWindowProps } from "./type";

export const ModalWindow = ({ children, onCloseHandler }: ModalWindowProps) => {
	const onClose = (e: MouseEvent<HTMLElement>) => {
		e.stopPropagation();

		if (e.target === e.currentTarget) onCloseHandler?.();
	};

	const createWrapper = () => (
		<StyledContainer>
			<StyledHeader>
				<CloseButton onClick={onClose} data-testid="modalCloseButton">
					<IconClose />
				</CloseButton>
			</StyledHeader>
			{children}
		</StyledContainer>
	);

	return ReactDOM.createPortal(
		<StyledCover data-testid="modalWindow" onClick={onClose}>
			{createWrapper()}
		</StyledCover>,
		document.getElementById("root") as HTMLElement
	);
};
