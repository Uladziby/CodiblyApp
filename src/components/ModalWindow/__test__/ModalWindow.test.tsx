/** @format */

import { render, RenderResult, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ModalWindow } from "../ModalWindow";

describe("modalWindow", () => {
	let renderResult: RenderResult;
	const prop = {
		children: <div>Portal</div>,
		onCloseHandler: jest.fn(),
	};
	let modalRoot = document.createElement("div");
	modalRoot.setAttribute("id", "root");

	beforeEach(() => {
		document.body.appendChild(modalRoot);
		renderResult = render(<ModalWindow {...prop} />);
	});

	it("should render correctly", () => {
		const element = renderResult.getByTestId("modalWindow");

		expect(element).toBeInTheDocument();
	});

	it("snapshot", () => {
		const element = renderResult.getByTestId("modalWindow");

		expect(element).toMatchSnapshot();
	});

	it("modal shows with close button", () => {
		const { getByRole } = screen;

		const closeButton = getByRole(/button/i);

		userEvent.click(closeButton);

		expect(prop.children).toEqual(<div>Portal</div>);
		expect(prop.onCloseHandler).toHaveBeenCalled();
	});

	it("should be mounted and unmounted", () => {
		renderResult.rerender(<ModalWindow {...prop} />);

		expect(modalRoot).not.toBeEmptyDOMElement();
		renderResult.unmount();
		expect(modalRoot).toBeEmptyDOMElement();
	});
});
