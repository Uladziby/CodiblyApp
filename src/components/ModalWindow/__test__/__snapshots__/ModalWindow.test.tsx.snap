// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modalWindow snapshot 1`] = `
<div
  class="sc-bcXHqe cuwjra"
  data-testid="modalWindow"
>
  <div
    class="sc-gswNZR eFkVuX"
  >
    <div
      class="sc-dkrFOg bJSqmQ"
    >
      <button
        class="sc-jSUZER bEElBe"
        data-testid="modalCloseButton"
      >
        <svg>
          clear.svg
        </svg>
      </button>
    </div>
    <div>
      Portal
    </div>
  </div>
</div>
`;
